cli/connect/history.go

```diff
@@ -0,0 +1,133 @@
+package connect
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/tarantool/tt/cli/util"
+)
+
+// commandHistory stores console command history.
+type commandHistory struct {
+	filepath    string
+	commands    []string
+	timestamps  []int64
+	maxCommands int
+}
+
+// parseHistoryCells extracts timestamps and commands from history lines.
+// If format doesn't contain timestamps, it sets the current timestamp
+// for each command.
+func parseHistoryCells(
+	lines []string) (commands []string, timestamps []int64) {
+	regex := regexp.MustCompile(`^#\d+$`)
+
+	// startPos is the first position of a timestamp.
+	startPos := -1
+	for i, line := range lines {
+		if regex.MatchString(line) {
+			startPos = i
+			break
+		}
+	}
+	timestamps = make([]int64, 0)
+	if startPos == -1 {
+		// Read one line per command.
+		// Set the current timestamp for each command.
+		commands = lines
+		timestamp := time.Now().Unix()
+		for range lines {
+			timestamps = append(timestamps, timestamp)
+		}
+		return
+	}
+
+	commands = make([]string, 0)
+	for startPos < len(lines) {
+		j := startPos + 1
+
+		// Move pointer to the next timestamp.
+		for j < len(lines) && !regex.MatchString(lines[j]) {
+			j++
+		}
+
+		// Extract the current timestamp.
+		timestamp, err := strconv.Atoi(lines[startPos][1:])
+
+		if j != startPos+1 && err == nil {
+			timestamps = append(timestamps,
+				int64(timestamp))
+			commands = append(commands,
+				strings.Join(lines[startPos+1:j], "\n"))
+		}
+		startPos = j
+	}
+
+	return
+}
+
+// load loads console history from the history file.
+func (history *commandHistory) load() error {
+	if history == nil {
+		return nil
+	}
+
+	rawLines, err := util.GetLastNLines(history.filepath, history.maxCommands)
+	if err != nil {
+		return err
+	}
+
+	history.commands, history.timestamps = parseHistoryCells(rawLines)
+	return nil
+}
+
+// appendCommand appends new command to the history.
+func (history *commandHistory) appendCommand(command string) {
+	if history == nil {
+		return
+	}
+
+	history.commands = append(history.commands, command)
+	history.timestamps = append(history.timestamps, time.Now().Unix())
+	if len(history.commands) > history.maxCommands {
+		history.commands = history.commands[1:]
+		history.timestamps = history.timestamps[1:]
+	}
+}
+
+// writeToFile writes console history to the file.
+func (history *commandHistory) writeToFile() error {
+	if history == nil {
```

Comment:
Here i did it for the next reason: if `newCommandHistory` returns an error, application should work correctly even without recording the console history. Because of this, ignore nil here.

Reply:
I understand it. I consider this functionality (history saving) as a potentially separate API, which does not know who calls it. And this is our console responsibility to decide if history saving is necessary. Yes, we can leave this check in our internal history saving implementation for now.

Reply:
Nice point, logic is separated now.